### ISSUE TITLE
Strange error (bug?) with lists of unions

### ISSUE THREAD

**Issue by Profpatsch** [dhall-lang/dhall-haskell#125]
```dhall
[ < A = 42 | B : Text >
, < A : Integer | B = "hello" > ]
: List < A: Integer | B: Text >
```

```
Error: Invalid input

(stdin):2:7: error: expected: "="
, < A : Integer | B = "hello" > ] 
      ^  
```

I’d expect it to work, similar to 

```haskell
data Foo = A Int | B Text
[ A 42, B "hello" ] :: [Foo]
```

Is this a bug? Or simply not possible in dhall?

**Comment by Gabriel439** [comment]
Not a bug.  In a union literal you need to put the value that you set at the beginning of the union, like this:

```haskell
< B = "hello" | A : Integer >
```

**Comment by Profpatsch** [comment]
Oh! That’s not clear from (or even mentioned in?) the tutorial.

Why that restriction?
If it’s to distinguish the value from the type, I’d argue that the difference between `=` and `:` is already so minuscule, it hardly even matters. I’d be in favor of letting the user keep the same ordering if she wants.

I’m kind of unhappy with how the value is nearly the same but requires restrictions.

```
< B = "hello" || A: Integer , C: Text > : < A: Integer | B: Text | C: Text >
or
< B = "hello" | A: Integer , C: Text > : < A: Integer, B: Text, C: Text >
even
```

, that would be a clearer distinction.

**Comment by Gabriel439** [comment]
So once #123 is merged I have another pull request that builds on top of that to let you set the value of a union anywhere inside it, such as:

```haskell
< A : Integer | B = True | C : Text >
```

I can explain the rationale behind why the separator is `|` ("pipe") for both union types and union literals

First, the separator is pipe for union types to make them resemble Haskell type declarations.  For example, the following Haskell type:

```haskell
data Example = Foo { bar :: Integer, baz :: Bool } | Qux { xyzzy :: Text }
```

... corresponds to this Dhall type:

```haskell
< Foo { bar : Integer, baz : Bool } | Qux { xyzzy : Text } >  
```

The reason why union literals also use `|` was because I originally planned to support the feature you requested (i.e. allowing the `=` anywhere in the union literal).  If you have that feature, then given a union type like:

```haskell
< A : Integer
| B : Bool
| C : Text
>
```

... then you can set any alternative by just replacing `: TheType` with `= TheValue`, like this:

```haskell
< A : Integer
| B = True
| C : Text
>
```

There are other reasons for using `|` as the separator, but those were the two strongest reasons.

The other reason was that I wanted to be able to easily visually distinguish union types and union literals from record types and record literals, especially once they get large.  You can tell that you are inside a large union type or union literal by the presence of the pipe (similar to how you can tell you are inside a large record type or record literal by the presence of the comma)

**Comment by Profpatsch** [comment]
Okay, so once the order of items doesn’t matter anymore I can accept the syntax for what it is. :)

It might make sense to improve pretty printing of options, though, to have the form you gave above:

```haskell
< A : Integer
| B : Bool
| C : Text
>
```
